### PR TITLE
Add GUC to enable/disable DML decompression

### PR DESCRIPTION
**Pull request by fabriziomello**
Disable-check: force-changelog-changed

### CODE REVIEW

File: src/guc.c [hunk 1]
```diff
@@ -269,6 +270,17 @@ _guc_init(void)
 							 NULL,
 							 NULL);
 
+	DefineCustomBoolVariable("timescaledb.enable_dml_decompression",
+							 "Enable DML decompression",
+							 "Enable DML decompression when querying hypertable",
```
Comment (sb230132) [inline src/guc.c:275]: It should be compressed hypertable.

File: tsl/test/expected/compression_conflicts.out [hunk 2]
```diff
@@ -493,3 +493,22 @@ WHERE hypertable_name = 'compressed_ht' ORDER BY chunk_name;
...
+\set ON_ERROR_STOP 0
+-- Should error because we disabled the DML decompression
+INSERT INTO compressed_ht VALUES ('2022-01-24 01:10:28.192199+05:30', '6', 0.876, 4.123, 'new insert row')
+  ON conflict(sensor_id, time)
+DO UPDATE SET sensor_id = excluded.sensor_id , name = 'ON CONFLICT DO UPDATE' RETURNING *;
+ERROR:  INSERT with ON CONFLICT DO UPDATE is not enabled on compressed chunks
+INSERT INTO compressed_ht VALUES ('2022-01-24 01:10:28.192199+05:30', '6', 0.876, 4.123, 'new insert row')
+  ON conflict(sensor_id, time)
+DO NOTHING;
+ERROR:  INSERT with ON CONFLICT DO UPDATE is not enabled on compressed chunks
```
Comment (sb230132) [inline tsl/test/expected/compression_conflicts.out:513]: I think you should have different message for each operation like:

"INSERT with ON CONFLICT DO UPDATE is not enabled on compressed chunks"  for ON CONFLICT DO UPDATE
"INSERT with ON CONFLICT DO NOTHING is not enabled on compressed chunks" for ON CONFLICT DO NOTHING

File: src/nodes/hypertable_modify.c [hunk 3]
```diff
@@ -752,6 +752,11 @@ ExecModifyTable(CustomScanState *cs_node, PlanState *pstate)
 	 */
 	if ((operation == CMD_DELETE || operation == CMD_UPDATE) && !ht_state->comp_chunks_processed)
 	{
+		if (!ts_guc_enable_dml_decompression)
```
Comment (sb230132) [inline src/nodes/hypertable_modify.c:755]: I think when ts_guc_enable_dml_decompression is set to false, even UPDATE/DELETE on regular hyperatable will fail. We don't want this behaviour.
IMO we should report error only when the chunk which is getting affected is compressed.

Best place to report error is in decompress_target_segments() before we call decompress_batches_for_update_delete().

**Review by sb230132 — COMMENTED** [review]
Thanks a lot for sharing the efforts.

File: src/nodes/hypertable_modify.c (hunk 3 above)
Comment (antekresic) [reply src/nodes/hypertable_modify.c:755]: I agree it seems like we have to push it down a bit until we can assert we are dealing with compressed chunks.

Comment (fabriziomello) [reply src/nodes/hypertable_modify.c]: You're correct. Fixed it and also added more tests creating new uncompressed chunk and executing UPDATE/DELETE on it... also testes UPDATE/DELETE with conditions that cross compressed and uncompressed chunks.

File: tsl/test/expected/compression_conflicts.out (hunk 2 above)
Comment (sb230132) [reply tsl/test/expected/compression_conflicts.out]: Sure.

File: tsl/test/sql/compression_update_delete.sql [hunk 4]
```diff
@@ -1068,3 +1070,73 @@ UPDATE join_test1 t1 SET value = t1.value + 1 FROM join_test1 t2 WHERE t2.time =
...
+            '2023-05-04 00:00:00-00'::timestamptz + INTERVAL '2 hours',
+            INTERVAL '1 hour') t;
+
+-- check chunk compression status
+SELECT chunk_name, is_compressed
+FROM timescaledb_information.chunks
+WHERE hypertable_name = 'sample_table'
+ORDER BY chunk_name;
+
+-- test for uncompressed and compressed chunks
```
Comment (sb230132) [inline tsl/test/sql/compression_update_delete.sql:1116]: You can ignore this comment. I think this test case is repetitive. We have lot of update/delete tests when this guc is turned ON.

**Review by sb230132 — APPROVED** [review]

File: tsl/test/sql/compression_update_delete.sql (hunk 4 above)
Comment (fabriziomello) [reply tsl/test/sql/compression_update_delete.sql:1116]: But this tests is to make sure we'll block if we touch both compressed and uncompressed chunks during UPDATE/DELETE execution

File: src/nodes/chunk_dispatch/chunk_insert_state.c [hunk 5]
```diff
@@ -585,6 +585,12 @@ ts_chunk_insert_state_create(const Chunk *chunk, ChunkDispatch *dispatch)
 	OnConflictAction onconflict_action = chunk_dispatch_get_on_conflict_action(dispatch);
 	ResultRelInfo *relinfo;
 
+	if (chunk->fd.compressed_chunk_id != INVALID_CHUNK_ID && !ts_guc_enable_dml_decompression)
```
Comment (antekresic) [inline src/nodes/chunk_dispatch/chunk_insert_state.c:588]: I think this also needs to get pushed down into decompress_batches_for_insert and block only if there are unique or primary constraints on the chunk.

Comment (fabriziomello) [reply src/nodes/chunk_dispatch/chunk_insert_state.c]: Fixed!

**Pull request merged**